src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs

```diff
@@ -308,15 +252,71 @@ public void GetCarbonIntensity_ConvertsMoerToGramsPerKwh(float lbsPerMwhEmission
+    [TestCase(new double[] { 0 }, null, TestName = "GetCarbonIntensity - for less than 2 data points, frequency is null for one data point ")]
+    [TestCase(new double[] { 300, 300 }, null, null, TestName = "GetCarbonIntensity - for multiple data points, frequency is null for all data points")]
+    [TestCase(new double[] { 500 }, 500, TestName = "GetCarbonIntensity - frequency is not null")]
+    [TestCase(new double[] { }, TestName = "GetCarbonIntensity - for zero data points, returns empty enumerable")]
+    public async Task GetCarbonIntensity_CalculatesDurationBasedOnFrequency(double[] durationValues, params int?[] frequencyValues)
     {
-        this.LocationSource.Setup(r => r.ToGeopositionLocationAsync(location)).Returns(Task.FromResult(location));
-        var latitude = location.Latitude.ToString() ?? throw new ArgumentNullException(String.Format("Could not find location"));
-        var longitude = location.Longitude.ToString() ?? throw new ArgumentNullException(String.Format("Could not find location"));
+        // Arrange
+        var startDate = this.DefaultDataStartTime;
+        var endDate = startDate.AddMinutes(10);
+        var emissionData = GenerateDataPoints(frequencyValues.Length);
+        for( int i = 0; i < frequencyValues.Length; i++)
+        {
+            emissionData[i].Frequency = frequencyValues[i];
+        }
+        
+        List<double> expectedDurationList = durationValues.ToList<double>();
+
+        this.WattTimeClient.Setup(w => w.GetDataAsync(
+            It.IsAny<BalancingAuthority>(),
+            It.IsAny<DateTimeOffset>(),
+            It.IsAny<DateTimeOffset>())
+        ).ReturnsAsync(() => emissionData);
+
+        // Act
+        var result = await this.DataSource.GetCarbonIntensityAsync(new List<Location>() { this.DefaultLocation }, startDate, endDate);
+
+        // Assert
+        List<double> actualDurationList = result.Select(e => e.Duration.TotalSeconds).ToList();
+        
+        CollectionAssert.AreEqual(expectedDurationList, actualDurationList);
+    }
 
-        this.WattTimeClient.Setup(w => w.GetBalancingAuthorityAsync(latitude, longitude)
-        ).ReturnsAsync(() => balancingAuthority);
+    private void MockBalancingAuthorityLocationMapping()
+    {
+        this.LocationSource.Setup(r => r.ToGeopositionLocationAsync(this.DefaultLocation)).Returns(Task.FromResult(this.DefaultLocation));
+        var latitude = this.DefaultLocation.Latitude.ToString();
+        var longitude = this.DefaultLocation.Longitude.ToString();
+
+        this.WattTimeClient.Setup(w => w.GetBalancingAuthorityAsync(latitude!, longitude!)
+        ).ReturnsAsync(() => this.DefaultBalancingAuthority);
+    }
+
+    private List<GridEmissionDataPoint> GenerateDataPoints(int numberOfDatapoints, float value = 10, DateTimeOffset startTime = default, int? frequency = 300)
```

Comment:
nit: Seems like `frequency` is not set anywhere. Suggestion is to keep it simple and not to pass args that are not used by anyone.

Reply:
frequency is used to set `Frequency` in setting the `GridEmissionDataPoint` object

Reply:
@pritipath in this case I think @juzuluag means that the argument is never used by any of the functions that call this method.  He is right and it is easy to add back if we write any unit tests that could benefit from passing this in.

I agree that as a matter of good practice we should not be writing "just-in-case" code, and instead wait until we need it.